xbmc/platform/linux/input/LibInputPointer.cpp

```diff
@@ -0,0 +1,143 @@
+
+#include "guilib/GraphicContext.h"
+#include "input/MouseStat.h"
+#include "utils/log.h"
+#include "windowing/WinEventsLinux.h"
+
+#include <algorithm>
+#include <libinput.h>
+#include <string.h>
+
+void CLibInputPointer::ProcessButton(libinput_event_pointer *e)
+{
+  const uint32_t b = libinput_event_pointer_get_button(e);
+  const bool pressed = libinput_event_pointer_get_button_state(e) == LIBINPUT_BUTTON_STATE_PRESSED;
+  unsigned char xbmc_button = 0;
+
+  switch (b)
+  {
+    case 0x110:
+      xbmc_button = XBMC_BUTTON_LEFT;
+      break;
+    case 0x111:
+      xbmc_button = XBMC_BUTTON_RIGHT;
+      break;
+    case 0x112:
+      xbmc_button = XBMC_BUTTON_MIDDLE;
+      break;
+    case 0x113:
+      xbmc_button = XBMC_BUTTON_X1;
+      break;
+    case 0x114:
+      xbmc_button = XBMC_BUTTON_X2;
+       break;
+    case 0x115:
+      xbmc_button = XBMC_BUTTON_X3;
+      break;
+    case 0x116:
+      xbmc_button = XBMC_BUTTON_X4;
+       break;
+    default:
+      break;
+  }
+
+  XBMC_Event event;
+  memset(&event, 0, sizeof(event));
+
+  event.type = pressed ? XBMC_MOUSEBUTTONDOWN : XBMC_MOUSEBUTTONUP;
+  event.button.button = xbmc_button;
+  event.button.x = static_cast<uint16_t>(m_pos.X);
+  event.button.y = static_cast<uint16_t>(m_pos.Y);
+
+  CLog::Log(LOGDEBUG, "CLibInputPointer::%s - event.type: %i", __FUNCTION__, event.type);
+  CLog::Log(LOGDEBUG, "CLibInputPointer::%s - event.button.button: %i", __FUNCTION__, event.button.button);
+  CLog::Log(LOGDEBUG, "CLibInputPointer::%s - event.button.x: %i", __FUNCTION__, event.button.x);
+  CLog::Log(LOGDEBUG, "CLibInputPointer::%s - event.button.y: %i", __FUNCTION__, event.button.y);
+
+  m_winEvents->MessagePush(&event);
+}
+
+void CLibInputPointer::ProcessMotion(libinput_event_pointer *e)
+{
+  const double dx = libinput_event_pointer_get_dx(e);
+  const double dy = libinput_event_pointer_get_dy(e);
+
+  m_pos.X += static_cast<int>(dx);
+  m_pos.Y += static_cast<int>(dy);
+
+  // limit the mouse to the screen width
+  m_pos.X = std::min(g_graphicsContext.GetWidth(), m_pos.X);
+  m_pos.X = std::max(0, m_pos.X);
+
+  // limit the mouse to the screen height
+  m_pos.Y = std::min(g_graphicsContext.GetHeight(), m_pos.Y);
+  m_pos.Y = std::max(0, m_pos.Y);
+
+  XBMC_Event event;
+  memset(&event, 0, sizeof(event));
+
+  event.type = XBMC_MOUSEMOTION;
+  event.motion.x = static_cast<uint16_t>(m_pos.X);
+  event.motion.y = static_cast<uint16_t>(m_pos.Y);
+
+  CLog::Log(LOGDEBUG, "CLibInputPointer::%s - event.type: %i", __FUNCTION__, event.type);
+  CLog::Log(LOGDEBUG, "CLibInputPointer::%s - event.motion.x: %i", __FUNCTION__, event.motion.x);
+  CLog::Log(LOGDEBUG, "CLibInputPointer::%s - event.motion.y: %i", __FUNCTION__, event.motion.y);
+
+  m_winEvents->MessagePush(&event);
+}
+
+void CLibInputPointer::ProcessAxis(libinput_event_pointer *e)
+{
+  unsigned char scroll = 0;
+  if (libinput_event_pointer_has_axis(e, LIBINPUT_POINTER_AXIS_SCROLL_VERTICAL))
+  {
```

Comment:
style: why not return early here if it doesn't have vertical and have the rest indented less?